store/sqlstore/channel_store_experimental.go

```diff
@@ -0,0 +1,843 @@
+	if err != nil {
+		return err
+	}
+
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels ON Channels
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP FUNCTION IF EXISTS channels_copy_to_public_channels
+		`); err != nil {
+			return err
+		}
+	} else if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_insert
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_update
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_delete
+		`); err != nil {
+			return err
+		}
+	} else if s.DriverName() == model.DATABASE_DRIVER_SQLITE {
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_insert
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_update_delete
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_update
+		`); err != nil {
+			return err
+		}
+		if _, err := transaction.Exec(`
+			DROP TRIGGER IF EXISTS trigger_channels_delete
+		`); err != nil {
+			return err
+		}
+	} else {
+		mlog.Critical("Failed to create trigger because of missing driver")
+		time.Sleep(time.Second)
+		os.Exit(EXIT_GENERIC_FAILURE)
+	}
+
+	if _, err := transaction.Exec(`
+		DROP TABLE IF EXISTS PublicChannels
+	`); err != nil {
+		return err
+	}
+
+	if err := transaction.Commit(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s SqlChannelStoreExperimental) CreateIndexesIfNotExists() {
+	s.SqlChannelStore.CreateIndexesIfNotExists()
+
+	if !s.IsExperimentalPublicChannelsMaterializationEnabled() {
+		return
+	}
+
+	s.CreateIndexIfNotExists("idx_publicchannels_team_id", "PublicChannels", "TeamId")
+	s.CreateIndexIfNotExists("idx_publicchannels_name", "PublicChannels", "Name")
+	s.CreateIndexIfNotExists("idx_publicchannels_delete_at", "PublicChannels", "DeleteAt")
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		s.CreateIndexIfNotExists("idx_publicchannels_name_lower", "PublicChannels", "lower(Name)")
+		s.CreateIndexIfNotExists("idx_publicchannels_displayname_lower", "PublicChannels", "lower(DisplayName)")
+	}
+	s.CreateFullTextIndexIfNotExists("idx_publicchannels_search_txt", "PublicChannels", "Name, DisplayName, Purpose")
+}
+
+func (s SqlChannelStoreExperimental) CreateTriggersIfNotExists() {
+	s.SqlChannelStore.CreateTriggersIfNotExists()
+
+	if !s.IsExperimentalPublicChannelsMaterializationEnabled() {
+		return
+	}
+
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		if !s.DoesTriggerExist("trigger_channels") {
+			transaction, err := s.GetMaster().Begin()
+			if err != nil {
+				mlog.Critical("Failed to create trigger function", mlog.Err(err))
```

Comment:
We could consider allowing this to return an error or something so that we don't need to repeat this error handling code so many times